File: leetcode/cpp/9_palindrome_number.cpp
#include <cmath>
#include <iostream>

class Solution {
 public:
  // another way : may use the reverse integer code
  // o(n)
  bool isPalindrome(int x) {
    if (x < 0) return false;
    int n = 0;
    int i, j;
    int temp = x;
    while (temp) {
      n++;
      temp = temp / 10;
    }
    // std::cout << n << std::endl;
    i = 0;
    j = n - 1;
    while (i < j) {
      if ((x / (int)pow(10, i)) % 10 != (x / (int)pow(10, j)) % 10) {
        break;
      } else {
        i++;
        j--;
      }
    }
    if (i < j)
      return false;
    else
      return true;
  }
};

int main(int argc, char *argv[]) {
  Solution s;
  std::cout << s.isPalindrome(12321) << std::endl;
  std::cout << s.isPalindrome(1221) << std::endl;
  std::cout << s.isPalindrome(22) << std::endl;
  return 0;
}

File: leetcode/go/src/leetcode/9_palindrome_number.go
package leetcode

func isPalindrome(x int) bool {
	if x < 0 {
		//var t int64 = -int64(x)
		//if (t > (1 << 31) - 1) {
		//    return false
		//}
		//return isPalindrome(-x)
		return false
	}
	i, t := 0, x
	for t > 0 {
		i++
		t = t / 10
	}
	for k := 0; k < (i / 2); k++ {
		if (x/pow(10, k))%10 != (x / pow(10, i-1-k) % 10) {
			return false
		}
	}
	return true
}

func pow(x int, n int) int {
	if n == 0 {
		return 1
	}
	if n%2 == 0 {
		res := pow(x, n/2)
		return res * res
	}
	return pow(x, n-1) * x

}
